Vision Partial Veneer

lukesneeringer opened the pull request:
This is the new surface with breaking changes.

This is an initial pass at the implementation of a "partial Veneer" for Vision.

It:

  * deprecates the entire manual layer
  * makes the GAPIC, with helpers, available at `require('@google-cloud/vision').v1()`.
  * provides the `annotateImage` and single-feature methods
  * has 100% test coverage on manual code

It does *not*:

  * ~provide a default partial Veneer on the versionless entry point (because it is taken by the deprecated manual layer)~ have backwards compatibility, at all, even remotely a little
  * [x] handle specifying a file handler for an image (needs to be done)
  * [x] provide documentation for the partial Veneer (needs to be done)
  * [x] have 100% test coverage on automated code (needs to be done)

Notes:

  * This introduces the use of ES6 syntax supported by Node 4.
  * This uses power-assert ~sinon and chai, which are not currently being used by other packages~

stephenplusplus commented:
> provide a default partial Veneer on the versionless entry point (because it is taken by the deprecated manual layer)

If the manual layer wasn't there, what would the ideal solution have been?

> This uses sinon and chai, which are not currently being used by other packages

Instinctively, I would think we should stick to our testing style in the rest of the repo. Is it just because it's easier, prettier, had some better features? We use sinon in places for easier stubbing, the chai part would be new.

jmdobry commented:
> provide a default partial Veneer on the versionless entry point (because it is taken by the deprecated manual layer)

In which version of `@google-cloud/vision` do you plan to export the partial gapic as the versionless entrypoint?

jmdobry commented:
> If the manual layer wasn't there, what would the ideal solution have been?

The ideal solution would be:

```js
module.exports = vision_v1;
module.exports.v1 = vision_v1;
// fictional other version
module.exports.v1beta1 = vision_v1beta1;
```

right?

stephenplusplus commented:
I think so, but I can see being explicit worth considering, as well, if we intend to carry multiple versions.

lukesneeringer commented:
> > If the manual layer wasn't there, what would the ideal solution have been?

> The ideal solution would be:

> ```js
module.exports = vision_v1;
module.exports.v1 = vision_v1;
// fictional other version
module.exports.v1beta1 = vision_v1beta1;
```

> right?

Precisely.

If you ask for a version, you get that version.
If you do not, you get the latest GA.

lukesneeringer commented:
> In which version of `@google-cloud/vision` do you plan to export the partial gapic as the versionless entrypoint?

I do not know; I was going to ask for feedback on that. :-) I did not realize I would run up against that problem until yesterday when I was writing the code.

There are a few options:

  * Move the old layer to a new namespace (backwards-incompatible change), and document that. Existing customers make a one-line change to "get working again" and then can go port their stuff.
  * The current form: For now, you need to use the version to get the new stuff. 
      * A hybrid of the above two: We do the current form and _also_ and old `manual` namespace; give people a month or so to make the one-line change, then they have a longer-deprecation period to update to the new surface.
  * Move fast and break things: Just wipe out the manual surface sooner in Node than in other languages. (In fairness, Node developers are likely _the most_ receptive to this option of any language community. :-))

lukesneeringer commented:
> Instinctively, I would think we should stick to our testing style in the rest of the repo. Is it just because it's easier, prettier, had some better features? We use sinon in places for easier stubbing, the chai part would be new.

I really needed sinon to do the mocks I wanted. Since you are using it elsewhere, I assume this is not interesting.

I ended up adding chai at the last minute because `assert` was not giving me a useful error message for a particular test failure, and I knew that chai would. I switched it, got the error I needed, and fixed the test. Once I fixed the test, I did not bother to change it back. On the one hand, I could totally change it back now; on the other hand, I prefer more useful error messages for future test failures. I could go either way on this.

lukesneeringer commented:
@stephenplusplus I would also like your thoughts on my use of ES6 syntax. Is that something you are okay with beginning to use now that we dropped Node 0.12 support, or would you prefer have the repo remain ES5 for the forseeable future?

stephenplusplus commented:
Move fast and break things.

I also don't feel strongly about chai or no chai, if it helps us. My concern would be adding another tool to learn that contributors have to face.

ES6 syntax is great.

I'll do a deeper dive soon.

jmdobry commented:
I agree, move fast and break things. It's better to remove the manual code pre-1.0 than to have to do a post-1.0 major version bump.

jmdobry commented:
> I prefer more useful error messages

The most useful error message come not from using chai, but by using [power-assert](https://github.com/power-assert-js/power-assert).

lukesneeringer commented:
> I agree, move fast and break things. It's better to remove the manual code pre-1.0 than to have to do a post-1.0 major version bump.

These are not the only options, though. We are not going 1.0 when we release this. We could deprecate the manual layer, and still remove it pre-1.0.

(Nonetheless, I am okay with "move fast and break things" in Node. I am not okay with it in Python and Ruby.)

lukesneeringer commented:
> The most useful error message come not from using chai, but by using power-assert.

Oh hey, I like that way better. (That is similar to py.test in Python.)
I will swap to that later.

lukesneeringer commented:
I moved over to power-assert. Holding off on scrapping the manual layer until that can have more discussion.

stephenplusplus commented on packages/vision/package.json:
```diff
@@ -70,10 +70,40 @@
+    "sinon": "^2.2.0",
     "tmp": "^0.0.31"
   },
+  "nyc": {
```
Maybe this can go into a .nycrc in the root directory and have the same effect. Or possibly better, we have a npm run-script, that is a small wrapper around the current `cover` task, that takes the module to test as an argument.

lukesneeringer replied:
I took this from the existing root `package.json`.

stephenplusplus replied:
The goal would be to not have duplicate places of configuration.

lukesneeringer replied:
I see.

jmdobry commented on packages/vision/src/helpers.js:
```diff
@@ -101,6 +106,19 @@ module.exports = apiVersion => {
+    // it to content.
+    if (request.image.source && request.image.source.filename) {
+      request.image = {
+        content: fs.readFileSync(request.image.source.filename),
```
We should defer to performing I/O asynchronously as much as possible, as this will block the user's whole app.

stephenplusplus commented on packages/vision/src/helpers.js:
```diff
@@ -0,0 +1,197 @@
+    });
+
+    // Assign the single-feature method to the `methods` object.
+    methods[methodName] = promisify(_createSingleFeatureMethod(featureValue));
```
Using a loop to define methods is a problem for the docs. I'm not sure if this will work, we might have to actually write out the methods in order to get them into the docs.

swcloud replied:
@stephenplusplus @lukesneeringer Two questions

(1) We define methods.annotateImage here and it has its comments as documentation. Is jsdoc able to surface this documentation?

(2) If (1) is true, then can we surface each individual single feature method in such a way (instead of using a loop)

/**
 *  Documentation for faceDetection
*/
methods.faceDetection = promisify(_createSingleFeatureMethod(features['FACE_DETECTION']));

/**
 *  Documentation for logoDetection
*/
methods.logoDetection = promisify(_createSingleFeatureMethod(features['LOGO_DETECTION']));

stephenplusplus replied:
Yes, JSDocs will only work if there is a comment block for each method, above the method definition, as in your example.

swcloud replied:
OK. I will push a commit to have individual methods with their documentation instead of a loop.

stephenplusplus commented on packages/vision/src/helpers.js:
```diff
@@ -0,0 +1,197 @@
+   *   API call.
+   *
+   * @example
+   * var vision = require('@google-cloud/vision').v1();
```
Is it important to use the `v1()` here, or can we just use "require('@google-cloud/vision')()", since they are the same?

If we don't need the `v1()`, you can actually omit this line. Our docs script will always define and show the instantiation of "var moduleName = require('@google-cloud/module-name')` in the docs. This way, all methods are free to use the `moduleName` variable, without having to show how to instantiate the module in each block.

stephenplusplus commented on packages/vision/src/helpers.js:
```diff
@@ -0,0 +1,197 @@
+   *   image: {source: {image_uri: 'gs://path/to/image.jpg'}},
+   *   features: [],
+   * };
+   * client.annotateImage(request).then(response => {
```
vision.annotateImage*

stephenplusplus commented on packages/vision/src/index.js:
```diff
@@ -20,2097 +20,45 @@
+  // Create the image annotator client with the provided options.
+  var client = gapic.v1(opts).imageAnnotatorClient(opts);
+  if (is.undefined(client.annotateImage)) {
+    Object.assign(client.constructor.prototype, helpers('v1'));
```
Like the feature methods defined inside the helpers.js file, this way of programmatically defining methods is a problem for the docs. Off hand, I'm not exactly sure how to solve for this.

stephenplusplus replied:
I resolved this part in my commit.

lukesneeringer commented:
The coverage change seems to actually be because so much code was removed as to materially affect the overall total. The Vision coverage is 100%.

lukesneeringer commented:
@stephenplusplus @jmdobry The ML folks want this delivered reasonably soon. Can you give it another look and see if everything looks good?

In particular, be looking at the _pattern_, since future GAPIC and partial Veneer releases will follow it.

stephenplusplus commented on package.json:
```diff
@@ -22,9 +23,11 @@
     "propprop": "^0.3.1",
     "semver": "^5.3.0",
     "shelljs": "^0.7.3",
+    "sinon": "^2.2.0",
```
I don't think this is being used.

stephenplusplus commented:
@lukesneeringer I'm okay with it overall. The only issue is the docs (https://github.com/GoogleCloudPlatform/google-cloud-node/pull/2298/files/6f3caf64a8887189591b4279e5370beb178e189b#diff-ef8f4754e676d8ad6a326bd2f9bb1d28, https://github.com/GoogleCloudPlatform/google-cloud-node/pull/2298/files/6f3caf64a8887189591b4279e5370beb178e189b#diff-99cc4574db14fc26a3c7bf0b5e2e6ca8).

stephenplusplus commented:
I added a commit which resolved one of the docs issues -- how to merge the methods in `helpers.js` with the `visionV1` definition.

![](https://user-images.githubusercontent.com/723048/26993699-5b9f7612-4d32-11e7-9729-30f872a8a5cd.png)

We are still stuck on this however: https://github.com/GoogleCloudPlatform/google-cloud-node/pull/2298/files/6f3caf64a8887189591b4279e5370beb178e189b#r121120113

swcloud commented:
@stephenplusplus @jmdobry Do you think this PR is able to be merged? DPE needs to work based on the partial veneer implementation as soon as possible.

swcloud commented:
@lukesneeringer @stephenplusplus Please take a look at my latest commit https://github.com/GoogleCloudPlatform/google-cloud-node/pull/2298/commits/ba50074a02acf945857d3c9ad4caf1c244926ee1

stephenplusplus commented:
LGTM.

stephenplusplus commented on packages/vision/package.json:
```diff
@@ -1,42 +1,16 @@
-    }
-  ],
-  "main": "./src/index.js",
+  "version": "0.7.1",
```
The version should remain the same as it is, 0.11.3. When we publish after the merge, it will become 0.12.0.

swcloud replied:
Done.

stephenplusplus commented on packages/vision/package.json:
```diff
@@ -1,42 +1,16 @@
-    }
-  ],
-  "main": "./src/index.js",
+  "version": "0.12.0",
```
Sorry for the confusion-- this should not be edited in a PR. It is bumped during our release process.

stephenplusplus commented:
@lukesneeringer close this, since this is going in #2421?

stephenplusplus commented:
Made some changes.

Docs: http://stephenplusplus.github.io/gcloud-node/#/docs/vision/master/vision

lukesneeringer commented:
Looks good to me. Thanks @stephenplusplus

julien-c commented:
I'm using the Vision API to annotate an image with several features at the same time.

Unless I'm missing something obvious, the new `0.12` SDK makes it pretty hard to do this (I had to read the code to find the feature type enumeration)

Happy to open a new issue if that makes sense.
